omegaconf/built_in_resolvers.py

```diff
@@ -0,0 +1,71 @@
+import os
+import warnings
+from typing import Any, Dict, List, Optional
+
+from ._utils import _DEFAULT_MARKER_, _get_value, decode_primitive
+from .base import Container
+from .errors import ValidationError
+from .grammar_parser import parse
+
+
+def decode(expr: Optional[str], _parent_: Container) -> Any:
+    """
+    Parse and evaluate `expr` according to the `singleElement` rule of the grammar.
+
+    If `expr` is `None`, then return `None`.
+    """
+    if expr is None:
+        return None
+
+    if not isinstance(expr, str):
+        raise TypeError(
+            f"`oc.decode` can only take strings or None as input, "
+            f"but `{expr}` is of type {type(expr).__name__}"
+        )
+
+    parse_tree = parse(expr, parser_rule="singleElement", lexer_mode="VALUE_MODE")
+    val = _parent_.resolve_parse_tree(parse_tree)
+    return _get_value(val)
+
+
+def dict_keys(in_dict: Dict[Any, Any]) -> List[Any]:
+    return list(in_dict.keys())
+
+
+def dict_values(in_dict: Dict[Any, Any]) -> List[Any]:
+    return list(in_dict.values())
```

Comment:
Handle case where input is of unsupported type.
I think we should support DictConfig and string to be used for selecting a node (which will then be verified as DictConfig). (possibly MutableSequence to support plain dict as well, but can we even get a plain dict here?)

Reply:
I was intentionally not checking the type to support any dict-like object. But ok to be stricter. Plain dicts are possible (typically they would come from outputs of other resolvers)

Reply:
You can check `isintance(in_dict, MutableMapping)` to allow for both.

> (typically they would come from outputs of other resolvers)

Yeah,I thought about it but aren't we now converting them to DictConfig automatically?

Reply:
> You can check `isintance(in_dict, MutableMapping)` to allow for both.

Yup, was actually going to use `Mapping` (since we're only reading data).

> > (typically they would come from outputs of other resolvers)
> 
> Yeah,I thought about it but aren't we now converting them to DictConfig automatically?

We're not (this is only at the very last step of interpolation resolution, not for intermediate computations, and anyway it's supposed to go away as mentioned [below](https://github.com/omry/omegaconf/pull/644#issuecomment-806180751))

Reply:
Also done in 872eb2d